// @modxclub/ui/src/Link/User/index.js
import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import Typography from "material-ui/Typography";


import Avatar from '../../Avatar';

import Link from '../';

import { withStyles } from 'material-ui/styles';
import Grid from 'material-ui/Grid';

const styles = {
  root: {
    display: 'inline-flex',
    alignItems: 'center',
    textDecoration: "none",
  },
  row: {
    display: 'inline-flex',
    alignItems: 'center',
    textDecoration: "none",
  },
};


export class UserLink extends Link {

  static propTypes = {
    classes: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired,
    withAvatar: PropTypes.bool.isRequired,
    showName: PropTypes.bool.isRequired,
  }


  static defaultProps = {
    withAvatar: true,
    showName: true,
  };


  render() {

    const {
      user,
      variant,
      withAvatar,
      classes,
      secondary,
      showName,
      ...other
    } = this.props;


    if (!user) {
      return null;
    }

    const {
      id,
      username,
      fullname,
      // lastname,
    } = user;

    const name = fullname || username;

    const url = `/profile/${username}/`;
    // const url = `/users/${id}`;


    let avatarLink = <Link
      key={id}
      to={url}
      href={url}
      title={fullname || username}
      {...other}
    >

      <Avatar
        user={user}
        size="small"
      />

    </Link>

    return showName ? (
      <Grid
        container
        className={classes.root}
      >

        <Grid
          item
        >
          {withAvatar && avatarLink || null}

        </Grid>

        <Grid
          item
          xs
        >

          <Link
            key={id}
            to={url}
            href={url}
            {...other}
          >

            {variant !== undefined ?
              <Typography
                variant="title"
              >
                {name}
                {this.props.position ? <span style={{ fontSize: '70%', fontStyle: 'italic' }}> - {this.props.position}</span> : null}
              </Typography>
              :
              name
            }

          </Link>

          {secondary}

        </Grid>

      </Grid>
    ) : avatarLink || null;
  }
}


export default withStyles(styles)(UserLink);
